src/ripple/app/misc/SHAMapStoreImp.cpp

```diff
@@ -446,23 +446,16 @@ SHAMapStoreImp::run()
                     ;
             }
 
-            std::string nextArchiveDir =
-                dbRotating_->getWritableBackend()->getName();
             lastRotated_ = validatedSeq;
-            std::shared_ptr<NodeStore::Backend> oldBackend;
-            {
-                std::lock_guard lock (dbRotating_->peekMutex());
-
-                state_db_.setState (SavedState {newBackend->getName(),
-                                                nextArchiveDir, lastRotated_});
-                clearCaches (validatedSeq);
-                oldBackend = dbRotating_->rotateBackends(
-                    std::move(newBackend),
-                    lock);
-            }
-            JLOG(journal_.warn()) << "finished rotation " << validatedSeq;
+            state_db_.setState(SavedState {
+                newBackend->getName(),
+                dbRotating_->getName(),
```

Comment:
Since we don't hold a lock here, we can't guarantee that the `dbRotating->getName()` is the same db as the one that's rotated out on line 456.

Maybe `rotateBackends` should return the name of the rotated db and we set the `state_db_` just after the rotation?

I'll also note that the old code had the same issue, so it's possible that the scenario above can't happen for some other reason, but I did want to point it out.

Reply:
Nice find, fixed.